File: LimSeoyeon/04_DFS/BOJ_1520_내리막길.cpp
#include <iostream>
using namespace std;

int arr[501][501];
int DP[501][501];

int dy[4] = { 0,0,-1,1 };
int dx[4] = { -1,1,0,0 };

int N, M;

int DFS(int y, int x) { //return값 경로 개수.
	if (DP[y][x] != -1) {
		return DP[y][x];
	}
	if (y == N - 1 && x == M - 1) {
		return 1;
	}
	DP[y][x] = 0;
	for (int i = 0; i < 4; i++) {
		int nextY = y + dy[i];
		int nextX = x + dx[i];
		if (nextY < 0 || nextX < 0 || nextY >= N || nextX >= M) continue;
		if (arr[nextY][nextX] < arr[y][x]) {
			DP[y][x] += DFS(nextY, nextX);
		}
	}

	return DP[y][x];
}

int main() {
	ios::sync_with_stdio(false);
	cin.tie(NULL);

	cin >> N >> M;
	for (int i = 0; i < N; i++) {
		for (int j = 0; j < M; j++) {
			cin >> arr[i][j];
			DP[i][j] = -1;
		}
	}
	cout << DFS(0, 0);

	//DP[i][j] = i,j에서 DP[N][M]까지 갈 수 있는 경로의 수 .
	//DFS만 돌리면 시간초과.

}

File: LimSeoyeon/04_DFS/BOJ_1520_내리막길.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class Main {
    static int[][] graph;
    static boolean[] visited;
    static int [][] DP;

    static int [] dx = {0,0,-1,1};
    static int [] dy = {-1,1,0,0};
    static int N;
    static int M;
    public static void main(String[] args) throws IOException {
        BufferedReader bf = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(bf.readLine());

        N = Integer.parseInt(st.nextToken());
         M = Integer.parseInt(st.nextToken());
        //배열 선언
        graph = new int[N][M];
        visited = new boolean[N*M];
        DP = new int[N][M];

        for(int i=0; i<N; i++){
            st = new StringTokenizer(bf.readLine());
            for(int j=0; j<M; j++){
                graph[i][j]= Integer.parseInt(st.nextToken());
                DP[i][j]=-1;
            }
        }
        System.out.println(dfs(0,0));
    }
    static int dfs(int y, int x) {
        if(DP[y][x]!=-1){
            return DP[y][x];
        }
        if(y==N-1 && x==M-1){
            return 1;
        }
        DP[y][x]=0;
        for(int i=0; i<4; i++){
            int nextY = y+dy[i];
            int nextX = x+dx[i];
            if(nextY<0|| nextX<0|| nextY>=N || nextX>=M) continue;
            if(graph[nextY][nextX]<graph[y][x]){
                DP[y][x] +=dfs(nextY,nextX);
            }
        }
        return DP[y][x];
    }
}
